### DISCUSSION TITLE
Hyperlinks lost while flattening PDF

### DISCUSSION THREAD

**Discussion by paritoshranjan** [pymupdf/PyMuPDF#2792] [Looking for help]
## Describe the bug (mandatory)
Hyperlinks are getting lost while flattening PDF.

I am using following code:

```
 doc = fitz.open("pdf-having-hyperlink.pdf")
 out = fitz.open()  

    for page in doc:
        w, h = page.rect.br  
        outpage = out.new_page(width=w, height=h)  
        pix = page.get_pixmap(dpi=300)  
        outpage.insert_image(page.rect, pixmap=pix)
    out.save(flattenedfilename, garbage=3, deflate=True)
    
    out.close();
    doc.close();
```
## To Reproduce (mandatory)
If a pdf contains hyperlink, and is flattened using the above code, the the hyperlinks do not work in the flattend pdf.

## Expected behavior (optional)
Hyperlinks should work even after flattening pdf.

## Your configuration (mandatory)
Windows
Python 3.11.5
PyMuPDF 1.23.5

**Comment by JorjMcKie** [comment]
> Hyperlinks should work even after flattening pdf.

Why?
The intention of flattening is to convert the PDF pages to images with no further properties.

If you nevertheless want links, then create your own code: extract the links from the source page and insert them on the page just created.

**Comment by paritoshranjan** [comment]
Would it be possible to insert link at the same location? Can you guide a bit with which methods to use?

**Reply by JorjMcKie** [reply]
The following should work rightaway:
```python
 doc = fitz.open("pdf-having-hyperlink.pdf")
 out = fitz.open()  

    for page in doc:
        w, h = page.rect.br  
        outpage = out.new_page(width=w, height=h)  
        pix = page.get_pixmap(dpi=300)  
        outpage.insert_image(page.rect, pixmap=pix)
        for link in page.get_links():
            outpage.insert_link(link)
    out.save(flattenedfilename, garbage=3, deflate=True)
    
    out.close()
    doc.close()
```
The above will work for all links to external resources just like that.
Also for internal links if the dictionary key `link["kind"] == fitz.LINK_GOTO`.
More difficult if `link["kind"] == 4` is a named destination.
Then the given name must be resolved to provide a page number and a point on that page.
Such a link looks like `{'kind': 4, 'xref': 1247, 'from': Rect(71.00399780273438, 243.5269775390625, 179.33700561523438, 254.35601806640625), 'name': 'nameddest=chapter.1', 'id': ''}`.
The name `chapter.1` must be resolved via `names["chapter.1"]` look at this snippet:
```python
import fitz_new as fitz  # the new PyMuPDF implemention must be used
doc=fitz.open(doc.name)
doc
Document('pymupdf.pdf')
names = doc.resolve_names()  # resolve all symbolic names in the PDF
names["chapter.1"]
{'page': 6, 'to': (72.0, 720.0), 'zoom': 0.0}
```
The return dictionary can then be used to make a `LINK_GOTO` compatible dictionary which can be inerted in the same way.
